torch/_inductor/comms.py

```diff
@@ -0,0 +1,262 @@
+from collections import defaultdict
+from typing import List
+
+import torch.fx as fx
+from . import ir, scheduler
+from .analysis import get_runtime_snode
+from .dependencies import WeakDep
+from .utils import is_local, print2
+
+
+# Used to ensure that iterating over a set is deterministic
+def tuple_sorted(x):
+    return sorted(tuple(x), key=lambda x: x.name)
+
+
+def sink_waits(result: List[fx.Node]) -> List[fx.Node]:
+    """
+    Greedily moves waits as late as possible (i.e. until we reach an use). Optimal in terms of
+    communication overlap.
+    """
+    new_result = []
+    cur_waits = set()
+    for node in result:
+        if isinstance(node.meta["fusion_meta"].snode.node, ir.Wait):
+            cur_waits.add(node)
+        else:
+            for wait in tuple_sorted(cur_waits):
+                if node in wait.users:
+                    new_result.append(wait)
+                    cur_waits.remove(wait)
+
+            new_result.append(node)
+    for node in tuple_sorted(cur_waits):
+        new_result.append(node)
+    return new_result
```

Comment:
Is converting to FX actually doing anything here?  You aren't actually modifying the FX graph (the FX linked list will contain the old order).

Reply:
Converting it from SchedulerNodes to a FX graph is just used to make it easier to work with the graph. We then just return some topological sort of the FX nodes as a list of nodes.

Reply:
IMO this just adds complexity by having two duplicate graph representations.  If there are gaps in the existing graph we should improve it rather than duplicating.

Reply:
Well, the existing graph isn't really a "graph" representation - there's no way of easily querying users or dependencies.

I think there's also just value in consistently working with the same graph representation between people writing passes on Aten FX graphs and people writing passes at the scheduler level.

Reply:
FX graph is missing typed edges between nodes, so can't represent scheduler graphs.  Let's just add `.users` and `.dependencies` to the existing graph, so the names are the same and reduce learning curve.